l3kernel/l3candidates.dtx

```diff
@@ -129,10 +129,11 @@
 %
 % \begin{function}[added = 2018-04-01]
 %   {
-%     \tl_build_put_left:Nn, \tl_build_put_left:Nx,
-%     \tl_build_gput_left:Nn, \tl_build_gput_left:Nx,
-%     \tl_build_put_right:Nn, \tl_build_put_right:Nx,
-%     \tl_build_gput_right:Nn, \tl_build_gput_right:Nx,
+%     \tl_build_put_left:Nn, \tl_build_put_left:Ne,
+%     \tl_build_gput_left:Nn, \tl_build_gput_left:Ne,
+%     \tl_build_put_right:Nn, \tl_build_put_right:Ne,
+%     \tl_build_put_right:Nx,
```

Comment:
Why is the `\tl_build_put_right:Nx` variant documented, but not `\tl_build_gput_right:Nx`? Imho, either document them both, or drop them both (even though there might be single packages in the wild using a then undocumented and deprecated function). Consistency might be key here. Even if both end up undocumented, `\cs_generate_variant:Nn` will still be able to produce them both, if someone so chooses.

Reply:
This really comes down to whether this step is a formal deprecation or simply a change of emphasise. We don't document formally deprecated functions but should cover everything else, even where we have non-consistent sets of functions. So it depends how others see the 'plan'.

Reply:
Note that the aim here is to reduce the `x`-type variants as e do need to be mindful of hash table usage: pruning over time is part of the 'plan'.

Reply:
Release 2022-01-12 claimed
https://github.com/latex3/latex3/blob/fedd9a9de989bf4a4156e1dae5e3f0f677520fa7/l3kernel/CHANGELOG.md?plain=1#L523-L525

Looks like another claim for function variants will appear soon.

Reply:
@muzimuzhi This one is tricky. On one hand, as long as the base function exists you can always use `\cs_generate_variant:Nn` to have the `x`-type versions. On the other, there is a risk of `Undefined control sequence`. I am hoping to see what others feel about this balance.

If the feeling is that we have to keep the `x`-type variants for ever (a reasonable position), then I can certainly go back over and re-add them all. There is then a question about hash table usage if one also wants a full set of `e`-type variants: personally I think we could live with that, but I know that this may not be the view of everyone.

Reply:
(We said we won't remove stuff, but didn't specify if we meant base functions or variants too - one way or the other we should tighten up)

Reply:
@Skillmon  I do see the argument for removing the docs for `x`-type variants, although `\exp_args:Nx` at least might have to stay (as it's used to create `x`-type variants themselves so is used by `l3kernel`). I got part-way through a change but think I need to see the overall feeling before I can move that forward: first lets see what the general  view is on the desirability of a change at all.

Reply:
Roughly what is the current "hash table usage" of expl3?

Reply:
@car222222 Building the format for pdfLaTeX in the `e-type` branch gives:
```
22938 strings of total length 430829
1916723 memory locations dumped; current usage is 204&347958
21242 multiletter control sequences
```
which is as close as I can come to an estimate.

Reply:
Thus, at an order of magnitude of 21242 (or even 10^5, the hash size does not seem to be a problem.  

At least not when compared with the per document requirements of a large document processed by latex-lab!

Reply:
OK, I'll adjust to keep the `x`-types but not document, combining suggestions :)

Reply:
Extra info: a recent (2023) setup of pdftex+miktex seems to provide the following size of hash-table: 15000+200000.

So running with expl3 plus a few "heavyweight packages" will typically require over 30% of this total before adding anything from the document itself (such as 10 or even 100 per page for tagging, etc.)

But note that most of this will now be in the "non-hashed" part of the table so access will be slower.

Reply:
Done: only costs around 90 csnames

Reply:
A mere drop in the ocean!

Reply:
@muzimuzhi Hopefully you prefer the updated version :)

Reply:
Personally I can live with one-time dropping followed by a claim that no variants will be removed (only deprecated) from now on. Letting breakage plays the role of `e`-type promoter? But yes an instant no-variant-removal policy gains most backwards compatibility for class and package authors.

Reply:
@muzimuzhi I think the plan of adding `\exp_args:N..x` to the deprecation module will do that for the 'interested' programmer -they'll get the warning, end users won't. I'll look at that later today/tomorrow.